src/language_features/kak_lsp_custom.rs

```diff
@@ -0,0 +1,318 @@
+use crate::{
+    context::*, language_features::hover::editor_hover, markup::escape_brace, types::*, util::*,
+};
+use lsp_types::{request::*, *};
+use serde::Deserialize;
+use url::Url;
+
+/// Entry point for the `kak-lsp/next-or-previous-symbol` functionality
+///
+/// Essentially all this method is doing is making a DocumentSymbolRequest
+/// and then passing on the results to `process_results_for_next_or_prev()`.
+///
+/// This function's code is derived from `language_features::text_document_document_symbol()`
+/// and similar to it. At this stage we just need to issue a `DocumentSymbol` request.
+pub fn document_next_or_prev_symbol_request(
+    meta: EditorMeta,
+    editor_params: EditorParams,
+    ctx: &mut Context,
+) {
+    let req_params = DocumentSymbolParams {
+        text_document: TextDocumentIdentifier {
+            uri: Url::from_file_path(&meta.buffile).unwrap(),
+        },
+        partial_result_params: Default::default(),
+        work_done_progress_params: Default::default(),
+    };
+    ctx.call::<DocumentSymbolRequest, _>(
+        meta,
+        req_params,
+        move |ctx: &mut Context, meta, result| {
+            process_results_for_next_or_prev(meta, editor_params, result, ctx)
+        },
+    );
+}
+
+/// When this function is called we have got the results of DocumentSymbolRequest.
+/// This function does some minimal processing and then hands the data off to an
+/// appropriate function that will search for the next/previous symbol.
+///
+/// After that is done, we call `next_or_prev_response()` which does the job
+/// of actually sending the response back to Kakoune.
+///
+/// This function's code is derived from `document_symbol::editor_document_symbol()`.
+/// and its basic structure is similar to it.
+fn process_results_for_next_or_prev(
+    meta: EditorMeta,
+    editor_params: EditorParams,
+    result: Option<DocumentSymbolResponse>,
+    ctx: &mut Context,
+) {
+    let params = NextOrPrevSymbolParams::deserialize(editor_params).unwrap();
+    let hover = params.hover;
+    let maybe_details = match result {
+        Some(DocumentSymbolResponse::Flat(mut result)) => {
+            if result.is_empty() {
+                return;
```

Comment:
these are redundant but why not I guess

Reply:
The `if result.is_empty()` check is consistent with code in `document_symbol::editor_document_symbol()`

Though I don't think it is redundant because an external LSP is returning back our results and we should generally be defensive about the kind of data it returns.

Reply:
my bad, I missed that we only return in the empty case, sorry